src/index.js

```diff
@@ -255,56 +257,31 @@ export default class Undo {
    * @returns {Boolean} true if a block was deleted previously.
    */
   contentWasCopied(state, compState, index) {
-    return Object.keys(state[index].data).length === 0
-      && JSON.stringify(compState[index + 1]) !== JSON.stringify(state[index + 1]);
+    return (
+      Object.keys(state[index].data).length === 0
+      && JSON.stringify(compState[index + 1]) !== JSON.stringify(state[index + 1])
+    );
   }
 
   /**
    * Decreases the current position and update the respective block in the editor.
    */
   async undo() {
     if (this.canUndo()) {
-      const { index: nextIndex, state: nextState } = this.stack[this.position];
-
-      this.position -= 1;
       this.shouldSaveHistory = false;
-      let { index } = this.stack[this.position];
-      const { state, caretIndex } = this.stack[this.position];
+      const { state: lastState, caretIndex } = this.undoStack.pop();
 
-      this.onUpdate();
-      const blockCount = this.blocks.getBlocksCount();
+      // Add formatter to identify the type of modification
+      const jsonPatch = jsonpatchFormatter.format(lastState, this.basicData);
 
-      if (!state[index]) {
-        index -= 1;
-        this.stack[this.position].index = index;
-      }
+      // Add the Undo state, caret and inverse operation in the Redo undoStack
+      this.redoStack.push({ state: lastState, caretIndex, jsonPatch });
 
-      if (this.blockWasDeleted(state, nextState)) {
-        this.insertDeletedBlock(state, nextState, index);
-      } else if (this.contentWasCopied(state, nextState, index)) {
-        await this.blocks.render({ blocks: state });
-        this.caret.setToBlock(index, 'end');
-      } else if (index < nextIndex && this.blockWasSkipped(state, nextState)) {
-        await this.blocks.delete(nextIndex);
-        this.caret.setToBlock(index, "end");
-      } else if (blockCount > state.length) {
-        await this.blocks.render({ blocks: state });
-        this.setCaretIndex(index, caretIndex);
-      } else if (this.blockWasDropped(state, nextState)) {
-        await this.blocks.render({ blocks: state });
-        this.caret.setToBlock(index, 'end');
-      } else if (this.contentChangedInNoFocusBlock(index, nextIndex)) {
-        const { id } = this.blocks.getBlockByIndex(nextIndex);
-
-        await this.blocks.update(id, state[nextIndex].data);
-        this.setCaretIndex(index, caretIndex);
-      }
+      // To build the previous state of 'basicData', removing the changes
+      // specified in 'lastState'
+      unpatch(this.basicData, lastState);
```

Comment:
here the **unpatch** function changes `this.basicData`  by reference right? I mean, **lastState** inverse operation is directly applied to `this.basicData` correct?

Reply:
Yes, the **patch** and **unpatch** operations act directly on the `this.basicData`, so we don't need to do an assignment